test/battle/move_effect/protect.c

```diff
@@ -0,0 +1,369 @@
+        PARAMETRIZE {protectMove = protectMoves[j]; usedMove = MOVE_WATER_GUN; }
+    }
+
+    GIVEN {
+        PLAYER(SPECIES_WOBBUFFET);
+        OPPONENT(SPECIES_WOBBUFFET);
+    } WHEN {
+        TURN { MOVE(opponent, protectMove); MOVE(player, usedMove); }
+        TURN {}
+    } SCENE {
+        ANIMATION(ANIM_TYPE_MOVE, protectMove, opponent);
+        MESSAGE("Foe Wobbuffet protected itself!");
+        NOT ANIMATION(ANIM_TYPE_MOVE, usedMove, player);
+        MESSAGE("Foe Wobbuffet protected itself!");
+        if (usedMove == MOVE_LEER) {
+            NOT ANIMATION(ANIM_TYPE_GENERAL, B_ANIM_STATS_CHANGE, opponent);
+        } else {
+            NOT HP_BAR(opponent);
+        }
+    }
+}
+
+SINGLE_BATTLE_TEST("King's Shield / Silk Trap / Obstruct protect from damaging moves only and lower Atk / Spd / Def by 1 / 1 / 2 for contact moves")
+{
+    u32 j;
+    static const u16 protectMoves[][3] = {{MOVE_KINGS_SHIELD, STAT_ATK, 1}, {MOVE_SILK_TRAP, STAT_SPEED, 1}, {MOVE_OBSTRUCT, STAT_DEF, 2}};
+    u16 protectMove = MOVE_NONE;
+    u16 usedMove = MOVE_NONE;
+    u16 statId = 0, lowersBy = 0;
+
+    for (j = 0; j < ARRAY_COUNT(protectMoves); j++)
+    {
+        PARAMETRIZE {protectMove = protectMoves[j][0]; statId = protectMoves[j][1]; lowersBy = protectMoves[j][2]; usedMove = MOVE_TACKLE; }
+        PARAMETRIZE {protectMove = protectMoves[j][0]; usedMove = MOVE_LEER; }
+        PARAMETRIZE {protectMove = protectMoves[j][0]; usedMove = MOVE_WATER_GUN; }
+    }
+
+    GIVEN {
+        PLAYER(SPECIES_WOBBUFFET);
+        OPPONENT(SPECIES_WOBBUFFET);
+    } WHEN {
+        TURN { MOVE(opponent, protectMove); MOVE(player, usedMove); }
+        TURN {}
+    } SCENE {
+        ANIMATION(ANIM_TYPE_MOVE, protectMove, opponent);
+        MESSAGE("Foe Wobbuffet protected itself!");
+        if (usedMove == MOVE_LEER) {
+            ANIMATION(ANIM_TYPE_MOVE, usedMove, player);
+            ANIMATION(ANIM_TYPE_GENERAL, B_ANIM_STATS_CHANGE, opponent);
+            NOT MESSAGE("Foe Wobbuffet protected itself!");
+        } else {
+            NOT ANIMATION(ANIM_TYPE_MOVE, usedMove, player);
+            MESSAGE("Foe Wobbuffet protected itself!");
+            NOT HP_BAR(opponent);
+            if (usedMove == MOVE_TACKLE) {
+                ANIMATION(ANIM_TYPE_GENERAL, B_ANIM_STATS_CHANGE, player);
+                if (statId == STAT_ATK) {
+                    MESSAGE("Wobbuffet's Attack fell!");
+                } else if (statId == STAT_SPEED) {
+                    MESSAGE("Wobbuffet's Speed fell!");
+                } else if (statId == STAT_DEF) {
+                    if (lowersBy == 2) {
+                        MESSAGE("Wobbuffet's Defense harshly fell!");
+                    }
+                }
+            }
+        }
+    } THEN {
+        if (usedMove == MOVE_TACKLE) {
+            EXPECT_EQ(player->statStages[statId], DEFAULT_STAT_STAGE - lowersBy);
+        }
+    }
+}
+
+// Also checks if it can faint the mon.
```

Comment:
I would make a separate test for it.

Reply:
Personally not a fan of this since the test would be pretty much the same and I'd rather not duplicate code, especially if tests allow us to parametrize.